File: berlinclock/javascript/js/berlinclock.js
"use strict";

const Ticker = require("./ticker.js");
const Lightbar = require("./lightbar.js");

class BerlinClock {
  constructor(hour, minute) {
    const ticker = new Ticker();
    const bigHours = new Lightbar(4);
    const smallHours = new Lightbar(4);
    const bigMinutes = new Lightbar(11);
    const smallMinutes = new Lightbar(4);

    ticker.lightbar = smallMinutes;
    smallMinutes.next = bigMinutes;
    bigMinutes.next = smallHours;
    smallHours.next = bigHours;

    Object.assign(this, {
      triggerTicks(ticks) {
        for (let i = 0; i < ticks; i++) {
          ticker.tick();
        }
      },

      toString() {
        let str = "";

        bigHours.lights().forEach(it => (str += it ? "H" : "-"));
        smallHours.lights().forEach(it => (str += it ? "h" : "-"));
        bigMinutes.lights().forEach(it => (str += it ? "M" : "-"));
        smallMinutes.lights().forEach(it => (str += it ? "m" : "-"));

        return str;
      }
    });

    let seconds = 3600 * hour + 60 * minute;
    this.triggerTicks(seconds);
  }
}

module.exports = BerlinClock;

File: berlinclock/kotlin/src/main/kotlin/berlinclock/BerlinClock.kt
package berlinclock

class BerlinClock(val hours: Int = 0, val minutes: Int = 0) {

    private val largeHours = LightBar(4)
    private val smallHours = LightBar(4, largeHours)
    private val largeMinutes = LightBar(11, smallHours)
    private val smallMinutes = LightBar(4, largeMinutes)

    private val ticker = Ticker {
        if (smallHours.isFull() && largeHours.isFull()) {
            smallHours.reset()
            largeHours.reset()
        }

        smallMinutes.increment()
    }

    init {
        val seconds = 3600 * hours + 60 * minutes
        triggerTicks(seconds)
    }

    fun triggerTicks(ticks: Int) {
        repeat(ticks) { ticker.tick() }
    }

    override fun toString() = StringBuilder().apply {
        append(largeHours.toString().replace('*', 'H'))
        append(smallHours.toString().replace('*', 'h'))
        append(largeMinutes.toString().replace('*', 'M'))
        append(smallMinutes.toString().replace('*', 'm'))
    }.toString()
}